### ISSUE TITLE
private-users-chown changes the permissions of a great many directories to 777

### ISSUE THREAD

**Issue by georgyo** [systemd/systemd#3547]
### Submission type
- [X] Bug report
- [ ] Request for enhancement (RFE)

_NOTE: Do not submit anything other than bug reports or RFEs via the issue tracker!_
### systemd version the issue has been seen with

```
# systemctl --version
systemd 230
+PAM -AUDIT -SELINUX -IMA -APPARMOR +SMACK -SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD +IDN`
```
### Used distribution

Archlinux running on linode (xen shared kernel)

~~I believe archlinux still hasn't enabled kernel support for user namespaces, but the xen kernel running on linode does. I'll do more testing later today on bare metal archlinux.~~ Archlinux stock kernels do not support user namespaces. This effect was realized because the linode kernel does support user namespaces.
### In case of bug report: Expected behaviour you didn't see

The permissions of the folders modified by systemd-nspawn to be maintained, with only the uid:gid being modifed.
### In case of bug report: Unexpected behaviour you saw

Many directories were changed to 777, including the following /, /usr, /usr/bin, /usr/lib, /etc and many more. Presenting a huge security issue.
### In case of bug report: Steps to reproduce the problem
1. Have a machine in /var/lib/machines with done with a clean install. `mkdir /var/lib/machines/archtest && pacstrap -i -d -G /var/lib/machines/archtest`
2. Noticed the file permissions

```
# ls -lhva /var/lib/machines/archtest
total 16K
drwxr-xr-x 1 root root  122 Jun 15 16:48 .
...
drwxr-xr-x 1 root root 1.5K Jun 15 16:48 etc
...
drwxr-xr-x 1 root root   70 Jun 15 16:48 usr
...
```
1. Startup the container `/usr/bin/systemd-nspawn --quiet --boot --link-journal=try-guest --network-veth -U --settings=override --machine=archtest`
2. Enter the container and notice the changed permissions

```
# machinectl shell archtest    
Connected to machine archtest. Press ^] three times within 1s to exit session.
[root@archtest ~]# ls -lhva /
total 16K
drwxrwxrwx   1 root  root   122 Jun 15 16:48 .
...
drwxrwxrwx   1 root  root  1.5K Jun 15 16:51 etc
...
drwxrwxrwx   1 root  root    70 Jun 15 16:48 usr
...
```
- Many other directories are affected, but not all of them. It seems targeted. Though I can't see how this would ever be useful.
- Restarting the container after fixing the permissions does not cause permissions to change again, I believe this is because it is picking the same UID:GID range.
### Other notes

My friend got hit by this pretty hard, with almost 15 containers getting their permissions changed, when systemd updated and he didn't noticed that -U became the default and rebooted. Learning from his sorrows, I have been doing `getfacl -R > oldperms` and `setfacl --restore=oldperms`

**Comment by poettering** [comment]
hmpf, this would be a major bug.

**Comment by zellio** [comment]
I am that friend. This is a major problem. :+1:

**Comment by petercolberg** [comment]
I worked around this issue by shifting the owner/group manually:

``` bash
#!/bin/bash
offset=$(( $(shuf -i 8-28671 -n 1) << 16 ))
while read -r -d '' uid gid file; do
    chown -h "$(( ($uid & 0xffff) + $offset )):$(( ($gid & 0xffff) + $offset ))" "$file"
done < <(find /var/lib/machines/test -printf '%U %G %p\0')
```

The limits for the user namespace offset are taken from `src/nspawn/nspawn.c`:

``` c
/* Note that devpts's gid= parameter parses GIDs as signed values, hence we stay away from the upper half of the 32bit
 * UID range here */
#define UID_SHIFT_PICK_MIN ((uid_t) UINT32_C(0x00080000))
#define UID_SHIFT_PICK_MAX ((uid_t) UINT32_C(0x6FFF0000))
```

What is the reason for not using limits of `0x00010000` and `0x7FFF0000`, respectively?

**Comment by poettering** [comment]
> What is the reason for not using limits of 0x00010000 and 0x7FFF0000, respectively?

see the comment about this on the original PR: https://github.com/systemd/systemd/pull/3093#discussion_r60848413

**Comment by poettering** [comment]
Quite an embarassing bug. Fix waiting in #3599.
